Trabalho2/grafo.c

```diff
@@ -10,57 +10,412 @@
+    {
+        if(strcmp(g->nos[i]->nome_user, nomeU) == 0){
+            return g->nos[i];
+        }
+    }
+
+    return NULL;
 }
 
 
 
 
 grafo * criaGrafo(tabela_dispersao *td)
 {
-    return NULL;
+    if(td == NULL){
+        return NULL;
+    }
+
+    grafo* g = grafo_novo();
+
+    if(g == NULL){
+        return NULL;
+    }
+
+    elemento *elem;
+    no_grafo *origem, *dest;
+    int msg[2];
+    msg[0] = msg[1] = 0;
+
+    for(int i=0;i<td->tamanho;i++){
+        elem = td->elementos[i];
+        while(elem != NULL)
+        {   
+            origem = no_insere(g, elem->msg->remetente);
+            dest = no_insere(g, elem->msg->destinatario);
+            
+            ligacao2(td, elem->msg->remetente, elem->msg->destinatario, msg);
+
+            if(msg[0]>0){
+                if(origem == NULL){
+                    origem = encontra_no(g, elem->msg->remetente);
+                }
+                if(dest == NULL){
+                    dest = encontra_no(g, elem->msg->destinatario);
+                }
+                cria_ligacao(origem, dest, msg[0]);
+            }
+
+            if(msg[1]>0){
+                if(origem == NULL){
+                    origem = encontra_no(g, elem->msg->remetente);
+                }
+                if(dest == NULL){
+                    dest = encontra_no(g, elem->msg->destinatario);
+                }
+                cria_ligacao(dest, origem, msg[1]);
+            }
+
+            elem = elem->proximo;
+        }
+    }
+
+    return g;
 }
 
 
 
 no_grafo **lista_amigos(grafo *g, char *nomeU, int *n)
 {
-   return NULL;
+    if(g == NULL){
+        return NULL;
+    }
+
+    if(nomeU == NULL){
+        return NULL;
+    }
+
+    if(n == NULL){
+        return NULL;
+    }
+
+    no_grafo ** amigos = NULL, **amigo =  NULL;
+    no_grafo *no;
+    int amigo1 = 0, count = 0;
+    *n = 0;
+
+    if((no = encontra_no(g, nomeU)) == NULL){
+        return NULL;
+    }
+
+    for(int i=0;i<no->tamanho;i++)
+    {
+        if(no->ligacoes[i]->peso_ligacao >= 4){
+            amigo1 = 1;
+            amigo = (no_grafo **) realloc(amigo, sizeof(no_grafo)*(count+1));
+            amigo[count] = no->ligacoes[i]->destino;
+            count++;
+        }
+    }
+    
+    if(amigo1 == 1){
+        for(int i=0;i<count;i++)
+        {
+            for(int j=0;j<amigo[i]->tamanho;j++)
+            {
+                if(strcmp(amigo[i]->ligacoes[j]->destino->nome_user, nomeU) == 0){
+                    if(amigo[i]->ligacoes[j]->peso_ligacao >= 4){
+                        amigos = (no_grafo **) realloc(amigos, sizeof(no_grafo)*((*n)+1));
+                        if(amigos == NULL){
+                            return NULL;
+                        }
+                        amigos[*n] = amigo[i];
+                        (*n)++;
+                        break;
+                    }
+                }
+            }
+        }   
+    }
+
+    if((*n)>0){
+        return amigos;

```

Comment:
you are leaking candidatos_amigos

Reply:
I noticed, I already fixed ever problem about Memory Leaks.